design-proposals/klip-15-new-api-and-client.md

```diff
@@ -0,0 +1,189 @@
+clarify the server side implementation result in a cleaner implementation and reduced lines of code.
+* The current query streaming endpoints (chunked response and Websockets) will be retired.
+* Any current Jetty specific plugins (e.g. security plugins) will be migrated to Vert.x
+* Vert.x has great support for working with various different network protocols and has has [unrivalled performance/scalability](https://www.techempower.com/benchmarks/)
+characteristics for a JVM toolkit. It will also set us up well for a fully async / reactive internal threading model in the server that we
+should aim towards for the future.
+
+### Implementation plan
+
+There is a lot of work to do here and it should be split up into separate work packages rather than trying to complete it as a single
+monolithic piece of work. Some obvious partitions of the work would be:
+
+* Protocol implementation
+* Java client implementation. This can be developed using a fake server so there is no dependency on completing the server work first.
+* JavaScript client implementation. Again, this can be developed using a fake server
+* Server streaming implementation
+* Migration of existing endpoints to new server implementation
+* Migration of existing plugins to new server implementation.
+* Switchover from old API to new API
+
+We should retain the current API implementation until it's ready to be switched over to the new implementation. The new implementation can exist
+in the same codebase during development, and only after switchover will the old implementation be removed.
+That way we can ensure a seamless switchover as there is a working API at all times.
+
+## Test plan
+
+We will require unit/module level tests and integration tests for all the new or changed components as per standard best practice.
+
+## Documentation Updates
+
+* We will produce new guide(s) for the Java and JavaScript clients.
+* We will produce example applications showing how to use the client in a real app. E.g. using Spring Boot / Vert.x (Java) and Node.js (JavaScript)
+* The current documentation for the HTTP/REST API can remain more or less as-is with the removal of the problematic parts.
+* We will write up a document on the new wire protocol to encourage 3rd parties to implement new clients.
+* There may be some server side configuration changes due to the new server implementation.
+
+# Compatibility Implications
+
+The current chunked response query streaming endpoint will be removed so any users currently using that will have to upgrade.
+
+The current websockets query streaming endpoint will be removed so any users currently using that will have to upgrade.
+This endpoint is currently undocumented so it's not expected a lot of users are using it. It is used by C3 (?)
+so that will need to be migrated to the new API.
+
+## Performance Implications
+
+* Streaming query results with the new client and protocol should provide significantly better performance than
```

Comment:
in terms of end-end latency of delivery? can you please clarify what metric might get better with this change?

Reply:
I think there are a few reasons why we could expect performance to improve:

* The current PollingSubscription is inefficient as it polls the internal blocking queue and when drained schedules a timer to try again in 100ms. This means throughput is effectively capped at 100 msgs every 100ms, in other words 1000 msgs per sec (not very fast!). With a non polling approach we should have no such limit and I know from experience with similar systems that we should be able push millions of small messages per second over the wire with a simple protocol. Now clearly we are also limited here by how fast KS can deliver messages to us, and that's unlikely to be millions of messages second. But if it can deliver more than a 1000 per second we should expect an increase in throughput.
* Latency should also improve as we won't have the 100ms timer.
* Because we won't be blocking threads to execute transient queries, we should have a much better resource usage on the server allowing us to scale to larger number of queries with lower RAM.

Reply:
Also... encoding overhead of the protocol is pretty low as its so simple. It should be faster to parse than HTTP. But again we may not notice the difference unless we push a lot of traffic.

Reply:
Thanks for clarifying.. Seems like the throughput would improve.. If you can distill some of these into the KLIP, that would be great. 

+1 on @apurvam 's comment.. Bottleneck could be somewhere else as well.. (we were surprised for e.g when pull queries bottlenecked on ksql code and not streams) and it may take other projects to actually realize this benefit..